src/audio/module_adapter/module/volume/volume_ipc4.c

```diff
@@ -0,0 +1,1433 @@
+#include <rtos/alloc.h>
+#include <rtos/init.h>
+#include <sof/lib/cpu.h>
+#include <sof/lib/memory.h>
+#include <sof/lib/uuid.h>
+#include <sof/list.h>
+#include <sof/math/numbers.h>
+#include <sof/platform.h>
+#include <rtos/string.h>
+#include <sof/trace/trace.h>
+#include <sof/ut.h>
+#include <ipc/control.h>
+#include <ipc/stream.h>
+#include <ipc/topology.h>
+#include <user/trace.h>
+
+#include <errno.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+LOG_MODULE_REGISTER(volume, CONFIG_SOF_LOG_LEVEL);
+
+#include <sof/audio/volume.h>
+
+#if CONFIG_IPC_MAJOR_3
+/* b77e677e-5ff4-4188-af14-fba8bdbf8682 */
+DECLARE_SOF_RT_UUID("pga", volume_uuid, 0xb77e677e, 0x5ff4, 0x4188,
+		    0xaf, 0x14, 0xfb, 0xa8, 0xbd, 0xbf, 0x86, 0x82);
+#else
+/* these ids aligns windows driver requirement to support windows driver */
+/* 8a171323-94a3-4e1d-afe9-fe5dbaa4c393 */
+DECLARE_SOF_RT_UUID("pga", volume_uuid, 0x8a171323, 0x94a3, 0x4e1d,
+		    0xaf, 0xe9, 0xfe, 0x5d, 0xba, 0xa4, 0xc3, 0x93);
+
+/* 61bca9a8-18d0-4a18-8e7b-2639219804b7 */
+DECLARE_SOF_RT_UUID("gain", gain_uuid, 0x61bca9a8, 0x18d0, 0x4a18,
+		    0x8e, 0x7b, 0x26, 0x39, 0x21, 0x98, 0x04, 0xb7);
+
+DECLARE_TR_CTX(gain_tr, SOF_UUID(gain_uuid), LOG_LEVEL_INFO);
+#endif
+
+DECLARE_TR_CTX(volume_tr, SOF_UUID(volume_uuid), LOG_LEVEL_INFO);
+
+#if CONFIG_FORMAT_S16LE
+/**
+ * \brief Used to find nearest zero crossing frame for 16 bit format.
+ * \param[in,out] source audio stream buffer.
+ * \param[in] frames Number of frames.
+ * \param[in,out] prev_sum Previous sum of channel samples.
+ */
+static uint32_t vol_zc_get_s16(const struct audio_stream __sparse_cache *source,
+			       uint32_t frames, int64_t *prev_sum)
+{
+	uint32_t curr_frames = frames;
+	int32_t sum;
+	int16_t *x = audio_stream_get_rptr(source);
+	int bytes;
+	int nmax;
+	int i, j, n;
+	const int nch = audio_stream_get_channels(source);
+	int remaining_samples = frames * nch;
+
+	x = audio_stream_wrap(source, x + remaining_samples - 1); /* Go to last channel */
+	while (remaining_samples) {
+		bytes = audio_stream_rewind_bytes_without_wrap(source, x);
+		nmax = VOL_BYTES_TO_S16_SAMPLES(bytes) + 1;
+		n = MIN(nmax, remaining_samples);
+		for (i = 0; i < n; i += nch) {
+			sum = 0;
+			for (j = 0; j < nch; j++) {
+				sum += *x;
+				x--;
+			}
+
+			/* first sign change */
+			if ((sum ^ *prev_sum) < 0)
+				return curr_frames;
+
+			*prev_sum = sum;
+			curr_frames--;
+		}
+		remaining_samples -= n;
+		x = audio_stream_rewind_wrap(source, x);
+	}
+
+	/* sign change not detected, process all samples */
+	return frames;
+}
+
+#endif /* CONFIG_FORMAT_S16LE */
+
+#if CONFIG_FORMAT_S24LE
+/**
+ * \brief Used to find nearest zero crossing frame for 24 in 32 bit format.
+ * \param[in,out] source audio stream buffer.
+ * \param[in] frames Number of frames.
+ * \param[in,out] prev_sum Previous sum of channel samples.
+ */
+static uint32_t vol_zc_get_s24(const struct audio_stream __sparse_cache *source,
```

Comment:
These dont look ipc4 specific ? aren't the volume operations all IPC generic code ?